### PR TITLE
feat: add `java/tests_unitaires`

### PR DESCRIPTION
**Pull request by LukaMrt**

### CODE REVIEW

File: java/tests_unitaires/README.md [hunk 1]
```diff
@@ -0,0 +1,13 @@
+# Les tests unitaires en Java
+
```
Comment (innocenzi) [inline java/tests_unitaires/README.md:2]:
```suggestion

> Écrit par [Luka](https://github.com/LukaMrt)

```

Author name is missing here

File: README.md [hunk 2]
```diff
@@ -19,3 +19,4 @@
 ## Java
 
 - `📑` [Interfaces fonctionnelles - syntaxe, avantages et exemples](java/interfaces_fonctionnelles)
+- `📑` [Tests unitaires - Utilisation en java](java/tests_unitaires)
```
Comment (innocenzi) [inline README.md:22]:
```suggestion
- `📑` [Tests unitaires - utilisation en java](java/tests_unitaires)
``` 

No need to capitalize a letter after a hyphen

File: java/tests_unitaires/README.md [hunk 3]
```diff
@@ -0,0 +1,13 @@
...
+Cette fiche pratique porte sur la création de tests unitaires en java. Nous utiliserons les librairies Junit, Assertj et Mockito.
+
+Nous prendrons l'exemple d'une classe très simple, une calculatrice, pour créer des tests basés dessus.
+
+Nous verrons les tests classiques ainsi que les mocks.
+
```
Comment (innocenzi) [inline java/tests_unitaires/README.md:8]: I think the extra new lines are not necessary here.

File: java/tests_unitaires/README.md [hunk 4]
```diff
@@ -0,0 +1,13 @@
...
+Cette fiche pratique n'est pas destinée à apprendre ce que sont les tests unitaires, quelle est leur utilité et comment les faire.
```
Comment (innocenzi) [inline java/tests_unitaires/README.md:9]:
```suggestion
Cette fiche pratique n'est pas destinée à apprendre ce que sont les tests unitaires, mais à voir quelle est leur utilité et comment les faire.
```

File: java/tests_unitaires/README.md [hunk 5]
```diff
@@ -0,0 +1,13 @@
...
+
+## Table des matières
+
+- [Tests unitaires - Utilisation en java](fr/TESTS_UNITAIRES_JAVA.md)
```
Comment (innocenzi) [inline java/tests_unitaires/README.md:13]:
```suggestion
- [Tests unitaires - utilisation en java](fr/TESTS_UNITAIRES_JAVA.md)
```

File: java/tests_unitaires/fr/TESTS_UNITAIRES_JAVA.md [hunk 6]
```diff
@@ -0,0 +1,171 @@
...
+```
+
+Les imports sont les suivants :
+
+```java
+import org.junit.Test;
+import static org.assertj.core.api.Assertions.assertThat;
+```
+
+On importe de manière static la méthode ``asserThat`` qui nous permet d'avoir une ligne très clair et parlante :
```
Comment (innocenzi) [inline java/tests_unitaires/fr/TESTS_UNITAIRES_JAVA.md:115]:
```suggestion
On importe de manière statique la méthode ``asserThat`` qui nous permet d'avoir une ligne très clair et parlante :
```

File: java/tests_unitaires/fr/TESTS_UNITAIRES_JAVA.md [hunk 7]
```diff
@@ -0,0 +1,171 @@
...
+import static org.mockito.Mockito.when;
+```
+
+L'enchainement des méthodes ``when(...).thenReturn(...)`` permet de définir quel sera le retour de la méthode lorsqu'elle sera appeler. Cela nous permet de maîtriser le retour qui sera fait lors de l'appel à la méthode dans ``divide``.
+
+## Conclusion
+
+C'est la fin de cette fiche pratique. Nous avons vu comment créer des tests unitaires en java avec les librairies Junit, Assertj et Mockito. Cela nous a permis de mettre en place une bonne couverture de test pour notre classe Calculator et ainsi la prévenir des erreurs et des régressions.
+
+La totalité du code se trouve dans [le dossier code](../code)
```
Comment (innocenzi) [inline java/tests_unitaires/fr/TESTS_UNITAIRES_JAVA.md:170]:
```suggestion
La totalité du code se trouve dans [le dossier code](../code).
``` 

End-of-line dot is missing.

File: java/tests_unitaires/fr/TESTS_UNITAIRES_JAVA.md [hunk 8]
```diff
@@ -0,0 +1,171 @@
...
+```
+
+C'est une classe qui contient des méthodes très simples à tester, le but étant seulement de découvrir les tests en java.
+
+NumberProvider est une interface qui fourni simplement un nombre, on pourrait imaginer une implémentation par exemple fournissant un nombre aléatoire.
+L'utilité fondamentale de cette interface n'est pas forcément incroyable mais elle nous permettra d'utiliser des mocks.
+
+Les points à retenir de cette classe Calculator pour les tests sont :
+
+ - Le field ``numberProvider`` qui nous permettra de d'expérimenter les mocks.
```
Comment (Tran-Antoine) [inline java/tests_unitaires/fr/TESTS_UNITAIRES_JAVA.md:53]: petite typo

File: java/tests_unitaires/fr/TESTS_UNITAIRES_JAVA.md [hunk 9]
```diff
@@ -0,0 +1,171 @@
...
+    @Before
+    public void setUp() {
+        this.numberProvider = Mockito.mock(NumberProvider.class);
+        this.calculator = new Calculator(10, numberProvider);
+    }
+
+}
+```
+
+La méthode ``Mockito.mock(T)`` retourne une instance de ``T`` que l'on pourra manipuler pour maîtriser les retours des méthodes de ce mock et donc nos tests.
```
Comment (Tran-Antoine) [inline java/tests_unitaires/fr/TESTS_UNITAIRES_JAVA.md:80]: I believe the correct signature is `Mockito.mock(Class<T>)`, according to your usage

File: java/tests_unitaires/fr/TESTS_UNITAIRES_JAVA.md [hunk 10]
```diff
@@ -0,0 +1,171 @@
...
+public void divide_whenGive0_shouldThrowIllegalArgumentException() {
+
+    int b = 0;
+
+    ThrowableAssert.ThrowingCallable method = () -> this.calculator.divide(b);
+    assertThatIllegalArgumentException().isThrownBy(method);
+}
+```
+
+Encore une fois on importe de manière static la méthode ``assertThatIllegalArgument`` pour avoir une ligne claire.
```
Comment (Tran-Antoine) [inline java/tests_unitaires/fr/TESTS_UNITAIRES_JAVA.md:134]:
```suggestion
Encore une fois on importe de manière statique la méthode ``assertThatIllegalArgument`` pour avoir une ligne claire.
```

File: java/tests_unitaires/fr/TESTS_UNITAIRES_JAVA.md [hunk 11]
```diff
@@ -0,0 +1,171 @@
...
+public interface NumberProvider {
+
+    int provideNumber();
+
+}
+```
+
+C'est une classe qui contient des méthodes très simples à tester, le but étant seulement de découvrir les tests en java.
+
+NumberProvider est une interface qui fourni simplement un nombre, on pourrait imaginer une implémentation par exemple fournissant un nombre aléatoire.
```
Comment (Tran-Antoine) [inline java/tests_unitaires/fr/TESTS_UNITAIRES_JAVA.md:48]:
```suggestion
NumberProvider est une interface qui fournit simplement un nombre, on pourrait imaginer une implémentation par exemple fournissant un nombre aléatoire.
```

File: java/tests_unitaires/fr/TESTS_UNITAIRES_JAVA.md [hunk 12]
```diff
@@ -0,0 +1,171 @@
...
+}
+```
+
+On importe (encore) une méthode de manière static, ``when``, pour clarifier.
+
+```java
+import static org.mockito.Mockito.when;
+```
+
+L'enchainement des méthodes ``when(...).thenReturn(...)`` permet de définir quel sera le retour de la méthode lorsqu'elle sera appeler. Cela nous permet de maîtriser le retour qui sera fait lors de l'appel à la méthode dans ``divide``.
```
Comment (Tran-Antoine) [inline java/tests_unitaires/fr/TESTS_UNITAIRES_JAVA.md:164]:
```suggestion
L'enchainement des méthodes ``when(...).thenReturn(...)`` permet de définir quel sera le retour de la méthode lorsqu'elle sera appelée. Cela nous permet de maîtriser le retour qui sera fait lors de l'appel à la méthode dans ``divide``.
```

**Comment by Tran-Antoine** [comment]
I would suggest that you specify somewhere that this is **not** a course on how unit tests actually work and what they are, but rather a presentation of the different testing frameworks in Java. Maybe add a prerequisites section where you explicitly state that a good knowledge in unit testing is required beforehand?

**Pull request merged**